File: Coding Problems - Python/Graphs/Number of Provinces.py
class Solution:
    def findCircleNum(self, isConnected: List[List[int]]) -> int:
        def dfs(node):
            visited[node] = 1
            for i in adjLs[node]:
                if not visited[i]:
                    dfs(i)
        n = len(isConnected)
        visited = [0] * n
        cnt = 0
        adjLs = [[]for _ in range(n)]
        for i in range(n):
            for j in range(n):
                if i != j and isConnected[i][j] == 1:
                    adjLs[i].append(j)
                    adjLs[j].append(i)

        for i in range(n):
            if not visited[i]:
                cnt += 1
                dfs(i)
        return cnt


        

File: Coding_Problems_Java/Number of Provinces.java
class Solution {
    private void dfs(int node, int[][] adjLs, int[] visited){
        visited[node] = 1;
        for(int i : adjLs[node]){
            if(visited[i] == 0){
                dfs(i, adjLs, visited);
            }
        }
    }
    
    public int findCircleNum(int[][] isConnected) {
        int n = isConnected.length;
        int[][] adjLs = new int[n][n];
        int [] visited = new int[n];
        int cnt = 0;

        for(int i = 0; i < n; i++){
            for (int j = 0; j < n; j++){
                if(i != j && isConnected[i][j] == 1){
                    adjLs[i][j] = j;
                    adjLs[j][j] = i;
                }
            }
        }
        for(int i = 0; i < n; i++){
            if (visited[i] == 0){
                cnt += 1;
                dfs(i, adjLs, visited);
            }
        }
        return cnt;
    }
}
